x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/ResultsIndexUpgradeService.java

```diff
@@ -0,0 +1,516 @@
+                    removeReadOnlyBlock(client, indicesToUpgrade,
+                        ActionListener.wrap(
+                            rrob -> adjustAliases(client,
+                                indexNameAndAliasProvider.oldIndicesWithReadAliases(),
+                                indexNameAndAliasProvider.newReadIndicesWithReadAliases(),
+                                readAliasListener),
+                            rrobFailure -> {
+                                String msg = "Failed making old indices writable again so that aliases can be moved. "
+                                    + rrobFailure.getMessage();
+                                logger.error(msg, rrobFailure);
+                                listener.onFailure(new ElasticsearchException(msg, rrobFailure));
+                            })
+                    ),
+                error -> {
+                    logger.error("Failed to reindex old read-only indices", error);
+                    removeReadOnlyBlock(client, indicesToUpgrade, ActionListener.wrap(
+                        empty -> listener.onFailure(error),
+                        removeReadOnlyBlockError -> {
+                            String msg = "Failed making old indices read/write again after failing to reindex: " + error.getMessage();
+                            logger.error(msg, removeReadOnlyBlockError);
+                            listener.onFailure(new ElasticsearchException(msg, removeReadOnlyBlockError));
+                        }
+                    ));
+                }
+            );
+
+            // <4> Old indexes are now readOnly, Time to reindex
+            ActionListener<AcknowledgedResponse> readOnlyListener = ActionListener.wrap(
+                ack -> reindexOldReadIndicesToNewIndices(client, indexNameAndAliasProvider.needsReindex(), request, reindexListener),
+                listener::onFailure
+            );
+
+            // <3> Set old indices to readOnly
+            ActionListener<AcknowledgedResponse> writeAliasesMovedListener = ActionListener.wrap(
+                resp -> setReadOnlyBlock(client, indicesToUpgrade, readOnlyListener),
+                listener::onFailure
+            );
+
+            // <2> Move write index alias to new write indices
+            ActionListener<AcknowledgedResponse> createWriteIndicesAndSetReadAliasListener = ActionListener.wrap(
+                resp -> adjustAliases(client,
+                    indexNameAndAliasProvider.oldIndicesWithWriteAliases(),
+                    indexNameAndAliasProvider.newWriteIndicesWithWriteAliases(),
+                    writeAliasesMovedListener),
+                listener::onFailure
+            );
+
+            // <1> Create the new write indices and set the read aliases to include them
+            createNewWriteIndicesIfNecessary(client, metaData, indexNameAndAliasProvider.newWriteIndices(),
```

Comment:
The creation of the write indices, and adding them to the read alias are not done at the same time just in case the write indices already exist (from manual creation, or from a previous upgrade attempt), but the aliases still need set.